Add the inreg attribute to sreg when present.

hjyamauchi opened the pull request:
On Windows/AArch64, a different register is used between when an arugment is both inreg and sret (X0 or X1) and when it is just sret (X8) as the following comment indicates:

https://github.com/llvm/llvm-project/blob/46fe36a4295f05d5d3731762e31fc4e6e99863e9/llvm/lib/Target/AArch64/AArch64CallingConvention.td#L42

```
  // In AAPCS, an SRet is passed in X8, not X0 like a normal pointer parameter.
  // However, on windows, in some circumstances, the SRet is passed in X0 or X1
  // instead.  The presence of the inreg attribute indicates that SRet is
  // passed in the alternative register (X0 or X1), not X8:
  // - X0 for non-instance methods.
  // - X1 for instance methods.

  // The "sret" attribute identifies indirect returns.
  // The "inreg" attribute identifies non-aggregate types.
  // The position of the "sret" attribute identifies instance/non-instance
  // methods.
  // "sret" on argument 0 means non-instance methods.
  // "sret" on argument 1 means instance methods.

  CCIfInReg<CCIfType<[i64],
      CCIfSRet<CCIfType<[i64], CCAssignToReg<[X0, X1]>>>>>,

  CCIfSRet<CCIfType<[i64], CCAssignToReg<[X8]>>>,
```

So missing/dropping inreg can cause a codegen bug.

This is a partial fix for https://github.com/swiftlang/swift/issues/74866

hjyamauchi commented:
@swift-ci please test

egorzhdan reviewed (approved):
I'm not familiar with this attribute, but the change looks reasonable, thanks!

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -3562,6 +3562,11 @@ llvm::CallBase *IRBuilder::CreateCallOrInvoke(
+        llvm::AttrBuilder builder(func->getContext());
+        builder.addAttribute(llvm::Attribute::InReg);
+        attrs = attrs.addParamAttributes(func->getContext(), argIndex, builder);
+      }
```
Hmm.  I see that this is the established pattern we're using here, but it's clearly wrong and won't handle indirect calls correctly; we need to be setting these attributes based on the `CGFunctionInfo`.

hjyamauchi replied:
@rjmccall It's not obvious to me, can you elaborate on how to set these attributes based on the CGFunctionInfo?

rjmccall replied:
You want the attribute to already be present in the attributes returned by `fn.getAttributes()`, which ultimately means it should be added in `SignatureExpansion::expandExternalSignatureTypes` based on the same logic that Clang uses to do it.

hjyamauchi committed:
Add the inreg attribute to sreg when present.

On Windows/AArch64, a different register is used between when an
arugment is both inreg and sret (X0 or X1) and when it is just sret
(X8) as the following comment indicates:

https://github.com/llvm/llvm-project/blob/46fe36a4295f05d5d3731762e31fc4e6e99863e9/llvm/lib/Target/AArch64/AArch64CallingConvention.td#L42

```
  // In AAPCS, an SRet is passed in X8, not X0 like a normal pointer parameter.
  // However, on windows, in some circumstances, the SRet is passed in X0 or X1
  // instead.  The presence of the inreg attribute indicates that SRet is
  // passed in the alternative register (X0 or X1), not X8:
  // - X0 for non-instance methods.
  // - X1 for instance methods.

  // The "sret" attribute identifies indirect returns.
  // The "inreg" attribute identifies non-aggregate types.
  // The position of the "sret" attribute identifies instance/non-instance
  // methods.
  // "sret" on argument 0 means non-instance methods.
  // "sret" on argument 1 means instance methods.

  CCIfInReg<CCIfType<[i64],
        CCIfSRet<CCIfType<[i64], CCAssignToReg<[X0, X1]>>>>>,

  CCIfSRet<CCIfType<[i64], CCAssignToReg<[X8]>>>,
```

So missing/dropping inreg can cause a codegen bug.

This is a partial fix for #74866

hjyamauchi commented:
@swift-ci please test

rjmccall reviewed (approved):
Thank you, that looks right.

compnerd commented:
@swift-ci please test Windows platform

hjyamauchi commented:
@swift-ci please test Windows platform

hjyamauchi commented:
@egorzhdan `inreg` shouldn't be there for x86_64 but I suppose the param attributes could change on a rebranch. How do I locally reproduce this (which branches do the swift and llvm-project repo need to be sync'd to)? CC @compnerd

hjyamauchi commented:
The CI link isn't the correct one?
I see
```Failed Tests (1):
  Swift(windows-x86_64) :: Interop/Cxx/stdlib/use-std-vector.swift
```

hjyamauchi commented:
CC @rjmccall @compnerd @hyp This PR missed an additional missing inreg on sret and https://github.com/swiftlang/swift/pull/76324 should fix it. Please take a look.

hjyamauchi commented:
@egorzhdan I will take a look at the test failure.

hjyamauchi commented:
@egorzhdan This PR https://github.com/swiftlang/swift/pull/76373 should already fix the failure CC @hyp

egorzhdan commented:
That's right, thanks @hjyamauchi.
